openedx/core/djangoapps/credentials/tasks/v1/tasks.py

```diff
@@ -54,3 +75,288 @@ def send_grade_to_credentials(self, username, course_run_key, verified, letter_g
+        status = cert_info.get('status', None)
+
+        send_grade_if_interesting(
+            user,
+            grade.course_id,
+            mode,
+            status,
+            grade.letter_grade,
+            grade.percent_grade,
+            verbose=verbose
+        )
+
+
+def paged_query(queryset, delay, page_size):
+    """
+    A generator that iterates through a queryset but only resolves chunks of it at once, to avoid overwhelming memory
+    with a giant query. Also adds an optional delay between yields, to help with load.
+    """
+    count = queryset.count()
+    pages = int(math.ceil(count / page_size))
+
+    for page in range(pages):
+        page_start = page * page_size
+        page_end = page_start + page_size
+        subquery = queryset[page_start:page_end]
+
+        if delay and page:
+            time.sleep(delay)
+        index = 0
+        try:
+            for item in subquery.iterator():
+                index += 1
+                yield page_start + index, item
+        except OperationalError:
+            # When running the notify_credentials command locally there is an
+            # OperationalError thrown by MySQL when there are no more results
+            # available for the queryset iterator. This change catches that exception,
+            # checks state, and then logs according to that state. This code runs in
+            # production without issue. This changes allows for the code to be run
+            # locally without a separate code path.
+            if index == count:
+                logger.info('OperationalError Exception caught, all known results processed in paged_query')
+            else:
+                logger.warning('OperationalError Exception caught, it is possible some results were missed')
+            continue
+
+
+def log_dry_run(certs, grades):
+    """Give a preview of what certs/grades we will handle."""
+    logger.info("DRY-RUN: This task would have handled changes for...")
+    ITEMS_TO_SHOW = 10
+
+    logger.info(f"{certs.count()} Certificates:")
+    for cert in certs[:ITEMS_TO_SHOW]:
+        logger.info(f"   {certstr(cert)}")
+    if certs.count() > ITEMS_TO_SHOW:
+        logger.info(f"    (+ {certs.count() - ITEMS_TO_SHOW} more)")
+
+    logger.info(f"{grades.count()} Grades:")
+    for grade in grades[:ITEMS_TO_SHOW]:
+        logger.info(f"   {gradestr(grade)}")
+    if grades.count() > ITEMS_TO_SHOW:
+        logger.info(f"    (+ {grades.count() - ITEMS_TO_SHOW} more)")
+
+
+def certstr(cert):
+    return f'{cert.course_id} for user {cert.user.id}'
+
+
+def gradestr(grade):
+    return f'{grade.course_id} for user {grade.user_id}'
+
+
+# This has Credentials business logic that has bled into the LMS. But we want to filter here in order to
+# not flood our task queue with a bunch of signals. So we put up with it.
+def send_grade_if_interesting(user, course_run_key, mode, status, letter_grade, percent_grade, verbose=False):
```

Comment:
Moved this from credentials/signals.py to better avoid what would’ve been a circular import. Also, where this method should live is tricky! It’s neither a signal receiver nor a celery task. But Matt convinced me that it fits better here if you think of it as a _task wrapper_.

I didn’t change anything at all about this method, just moved it.

([See original import solution and comment here.](https://github.com/edx/edx-platform/pull/27322#discussion_r612699995))

Reply:
Is this used anywhere else in the codebase? Will moving it cause other issues?

Reply:
Aside from here, it’s only used in tests which I am fixing presently 💃